File: Day19/Part4/shortest_path_all_keys.py
# LC 864 — Shortest Path to Get All Keys
# BFS with bitmask of keys
from collections import deque
from typing import List

def shortestPathAllKeys(grid: List[str]) -> int:
    R, C = len(grid), len(grid[0])
    start = None
    total_keys = 0
    for i in range(R):
        for j in range(C):
            ch = grid[i][j]
            if ch == '@':
                start = (i, j)
            if 'a' <= ch <= 'f':
                total_keys = max(total_keys, ord(ch) - ord('a') + 1)
    all_mask = (1 << total_keys) - 1
    q = deque()
    visited = [[[False]*(1<<total_keys) for _ in range(C)] for __ in range(R)]
    sx, sy = start
    visited[sx][sy][0] = True
    q.append((sx, sy, 0, 0))  # x, y, mask, dist
    DIRS = [(1,0),(-1,0),(0,1),(0,-1)]
    while q:
        x, y, mask, dist = q.popleft()
        if mask == all_mask:
            return dist
        for dx, dy in DIRS:
            nx, ny = x+dx, y+dy
            if nx < 0 or ny < 0 or nx >= R or ny >= C:
                continue
            ch = grid[nx][ny]
            if ch == '#':
                continue
            nmask = mask
            if 'a' <= ch <= 'f':
                nmask |= 1 << (ord(ch) - ord('a'))
            if 'A' <= ch <= 'F':
                need = 1 << (ord(ch) - ord('A'))
                if (nmask & need) == 0:
                    continue
            if not visited[nx][ny][nmask]:
                visited[nx][ny][nmask] = True
                q.append((nx, ny, nmask, dist+1))
    return -1

if __name__ == "__main__":
    grid = ["@.a..","###.#","b.A.B"]
    print(shortestPathAllKeys(grid))  # 8

File: Day19/Part4/shortest_path_all_keys.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 864 — Shortest Path to Get All Keys
// BFS with state (x, y, mask)

int shortestPathAllKeys(vector<string>& grid){
    int R = grid.size(), C = grid[0].size();
    int sx=0, sy=0, totalKeys=0;
    for(int i=0;i<R;++i){
        for(int j=0;j<C;++j){
            char ch = grid[i][j];
            if (ch=='@'){ sx=i; sy=j; }
            if (ch>='a' && ch<='f') totalKeys = max(totalKeys, ch - 'a' + 1);
        }
    }
    int allMask = (1<<totalKeys) - 1;
    vector<vector<vector<char>>> vis(R, vector<vector<char>>(C, vector<char>(1<<totalKeys, 0)));
    queue<tuple<int,int,int,int>> q; // x,y,mask,dist
    vis[sx][sy][0]=1; q.emplace(sx,sy,0,0);
    int dx[4]={1,-1,0,0}; int dy[4]={0,0,1,-1};
    while(!q.empty()){
        auto [x,y,mask,dist] = q.front(); q.pop();
        if (mask == allMask) return dist;
        for(int k=0;k<4;++k){
            int nx=x+dx[k], ny=y+dy[k];
            if (nx<0||ny<0||nx>=R||ny>=C) continue;
            char ch = grid[nx][ny];
            if (ch=='#') continue;
            int nmask = mask;
            if (ch>='a' && ch<='f') nmask |= 1 << (ch - 'a');
            if (ch>='A' && ch<='F'){
                int need = 1 << (ch - 'A');
                if ((nmask & need) == 0) continue;
            }
            if (!vis[nx][ny][nmask]){
                vis[nx][ny][nmask]=1;
                q.emplace(nx,ny,nmask,dist+1);
            }
        }
    }
    return -1;
}

int main(){
    vector<string> grid = {"@.a..","###.#","b.A.B"};
    cout << shortestPathAllKeys(grid) << "\n"; // 8
    return 0;
}
